### PR TITLE
feat: keystone crypto hd key

### PR DESCRIPTION
**Pull request by iGroza**
Fixes #

## Proposed Changes

- import accounts by crypto HD key'

## Checklist

- [x] I have read the rules of [**contribution**](https://github.com/haqq-network/haqq-wallet/blob/main/docs/contribution.md#important-steps).

### CODE REVIEW

File: ios/Podfile.lock [hunk 1]
```diff
@@ -1252,6 +1252,6 @@ SPEC CHECKSUMS:
   Yoga: 3efc43e0d48686ce2e8c60f99d4e6bd349aff981
   YogaKit: f782866e155069a2cca2517aafea43200b01fd5a
 
-PODFILE CHECKSUM: e3b1f869e456e9b403ae2a25a2b47c9e8619ca5d
+PODFILE CHECKSUM: 47e53b72709a00bb405eabdba170afb65dd87bdb
 
-COCOAPODS: 1.14.3
+COCOAPODS: 1.13.0
```
Comment (vivalaakam) [inline ios/Podfile.lock:1257]: update pods

File: package.json [hunk 2]
```diff
@@ -43,7 +43,7 @@
     "@haqq/provider-web3-utils": "0.0.14",
     "@haqq/shared-react-native": "0.0.9",
     "@invertase/react-native-apple-authentication": "2.2.2",
-    "@keystonehq/bc-ur-registry-eth": "^0.19.1",
+    "@keystonehq/bc-ur-registry-eth": "^0.7.7",
```
Comment (vivalaakam) [inline package.json:46]: What is the reason for a downgrade?

Comment (ragozin-nikita) [inline package.json:46]: Please fix the version

File: src/helpers/provider-instance.ts [hunk 3]
```diff
@@ -101,6 +102,18 @@ export async function getProviderInstanceForWallet(
           }),
         );
         break;
+      case WalletType.keystone:
+        cache.set(
+          id,
+          new ProviderKeystoneReactNative({
+            cryptoHDKeyCBORHex: wallet.accountId!,
```
Comment (vivalaakam) [inline src/helpers/provider-instance.ts:109]: Maybe there exists a simpler name?

File: src/components/image-description-block.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  source,
+  style,
+  title,
+  description,
+  titleParams,
+  descriptionParams,
+  onPress,
+}: ImageDescriptionBlockProps) => {
+  return (
+    <View style={StyleSheet.flatten([styles.container, style])}>
```
Comment (ragozin-nikita) [inline src/components/image-description-block.tsx:39]: https://reactnative.dev/docs/stylesheet#flatten

Flatten will be recreated every render, so you need to memoize it
```suggestion
    <View style={StyleSheet.flatten([styles.container, style])}>
```

File: src/components/keystone/keystone-accounts.tsx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+          listRef.current.scrollToEnd({
+            animated: true,
+          });
+        }
+      });
+    }, [addresses]);
+
+    return (
+      <PopupContainer plain>
+        <FlatList
```
Comment (ragozin-nikita) [inline src/components/keystone/keystone-accounts.tsx:44]: Missing `keyExtractor` prop

File: src/models/wallet.ts [hunk 6]
```diff
@@ -268,7 +268,9 @@ class WalletStore implements MobXStoreFromRealm {
   }
 
   getById(id: string = '') {
-    return this.wallets.find(wallet => wallet.address === id) ?? null;
+    return (
+      this.wallets.find(wallet => wallet.address === id.toLowerCase()) ?? null
+    );
```
Comment (ragozin-nikita) [inline src/models/wallet.ts:273]: Prettier issue?

File: src/screens/DeviceStack/KeystoneStack/keystone-accounts.tsx [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+export const KeystoneAccountsScreen = memo(() => {
+  const navigation = useTypedNavigation<SignInStackParamList>();
+  const {cryptoHDKeyCBORHex, ...params} = useTypedRoute<
+    KeystoneStackParamList,
+    KeystoneStackRoutes.KeystoneAccounts
+  >().params;
+  const provider = useRef(
+    new ProviderKeystoneReactNative({
+      cryptoHDKeyCBORHex,
+      awaitForSign: Promise.resolve,
```
Comment (ragozin-nikita) [inline src/screens/DeviceStack/KeystoneStack/keystone-accounts.tsx:32]: Need implementation
```suggestion
      awaitForSign: Promise.resolve,
```

File: src/screens/DeviceStack/KeystoneStack/keystone-accounts.tsx [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    },
+    [addresses, loading],
+  );
+
+  useEffectAsync(async () => {
+    setLoading(true);
+    try {
+      await loadMore();
+    } catch (error) {
+      Logger.captureException(error, 'chooseAccount');
```
Comment (ragozin-nikita) [inline src/screens/DeviceStack/KeystoneStack/keystone-accounts.tsx:98]:
```suggestion
      Logger.captureException(error, 'keystone chooseAccount');
```

File: src/screens/DeviceStack/KeystoneStack/keystone-accounts.tsx [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+      setAddresses([...newAddresses]);
+      updateWalletsToCreate(newAddresses.filter(address => !!address.exists));
+    },
+    [addresses],
+  );
+
+  const onAdd = useCallback(async () => {
+    walletsToCreate.forEach(item => {
+      Wallet.create(item.name, item);
```
Comment (ragozin-nikita) [inline src/screens/DeviceStack/KeystoneStack/keystone-accounts.tsx:151]: Please add support for `isImported` field in `.create` method.
```suggestion
      Wallet.create(item.name, {...item, isImported: true});
```

Or you can update created wallet:
```
Wallet.update(item.name, {isImported: true})
```

File: src/helpers/provider-instance.ts (hunk 3 above)
Comment (iGroza) [reply src/helpers/provider-instance.ts:109]: fix in next feature, when add **crypto-account** CBOR support, now supported only **crypto-hd**

File: src/screens/DeviceStack/KeystoneStack/keystone-accounts.tsx [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+      //@ts-ignore
+      navigation.navigate(KeystoneStackRoutes.KeystoneFinish, params);
+    }
+  }, [
+    walletsToCreate,
+    params,
+    navigation,
+    provider.current,
+    cryptoHDKeyCBORHex,
+  ]);
```
Comment (ragozin-nikita) [inline src/screens/DeviceStack/KeystoneStack/keystone-accounts.tsx:167]:
```suggestion
,app.onboarded
  ]);
```

File: src/screens/DeviceStack/KeystoneStack/keystone-finish.tsx [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+export const KeystoneFinishScreen = memo(() => {
+  const navigation = useTypedNavigation<KeystoneStackParamList>();
+  const onEnd = useCallback(() => {
+    if (app.onboarded) {
+      navigation.getParent()?.goBack();
+    } else {
+      app.onboarded = true;
+    }
+  }, [navigation]);
```
Comment (ragozin-nikita) [inline src/screens/DeviceStack/KeystoneStack/keystone-finish.tsx:20]:
```suggestion
  }, [navigation, app.onboarded]);
```

File: src/screens/DeviceStack/KeystoneStack/keystone-store-wallet.tsx [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+      ).length;
+
+      Logger.log('params', JSON.stringify(route?.params, null, 2));
+
+      actions.push(
+        Wallet.create(`${KEYSTONE_NAME} #${lastIndex}`, {
+          type: WalletType.keystone,
+          path: route.params.hdPath,
+          address: route.params.address,
+          accountId: route?.params?.cryptoHDKeyCBORHex,
```
Comment (ragozin-nikita) [inline src/screens/DeviceStack/KeystoneStack/keystone-store-wallet.tsx:48]:
```suggestion
          accountId: route?.params?.cryptoHDKeyCBORHex,
          isImported: true
```

**Review by ragozin-nikita — CHANGES_REQUESTED** [review]

**Pull request merged**